architecture/README.md

```diff
@@ -0,0 +1,654 @@
+
+#### Steps
+
+Pre-requisite steps to publishing data not in this workflow as we will not be changing that part of the process. The following steps follow
+on from a florence (DP internal user) publishes a collection that can contain 1 to many ons webpages or datasets by making a request to zebedee
+which does he publishing of new webpages or updates to existing pages and lastly triggering updates to sitewide search; which is where the flow begins.
+
+**1: Consume Kafka messages**
+
+On publishing a collection, consume kafka message
+
+```
+Datastore: kafka
+
+Topic: content-updated
+Record: {
+    "uri": string, // mandatory
+    "search_index": string, // mandatory
+    "data_type": string, // mandatory
+    "collection_id": string, // optional
+    "job_id": string // optional (used for reindex pipeline)
+}
+```
+
+The `search_index` field should be set to the search index alias, `ONS` for publishing new content. Only search reindexes will make use of the actual index name, e.g. `ons_<timestamp>`.
+
+**2: Retrieve Zebedee resource**
+
+Retrieve resource from Zebedee API endpoint based on the uri in kafka maessage.
+
+**3: Read JSON file from disc**
+
+Retrieve single document from zebedee content (files on disc).
+
+**4: Retrieve dataset resource**
+
+Retrieve resource from Dataset API endpoint based on the uri in kafka maessage.
+
+*Note: No parellisation with 2 as each of these are triggered by separate kafka messages*
+
+**5: Find latest version of dataset**
+
+Find one document from Mongo db representing the resource on given uri.
+
+**6: Produce event to search-data-import**
+
+```
+Datastore: kafka
+
+Topic: search-data-import
+Record: {
+    "data_type": string,
+    "job_id": string, // empty
+    "search_index": string, // Should use search alias ONS
+    "cdid": string,
+    "dataset_id": string,
+    "description": string,
+    "edition:: string,
+    "keywords": string,
+    "meta_description": string,
+    "release_date": string // date format: ISO8601 or strict_date_optional_time||epoch_millis to match existing docs in search?
+    "summary": string,
+    "title": string,
+    ... other fields we decide need to be in search
+}
+```
+
+See [full schema here](https://github.com/ONSdigital/dp-search-data-extractor/blob/develop/schema/schema.go#L25)
+
+**7: Consume event from search-data-import**
+
+Kafka message consumed by Search Data Importer. Documents received via consumption of kafka topic `search-data-import`, store documents in memory until 500 messages consumed or time from first message exceeded 5 seconds before making bulk request. 5 second limit will allow for the last set of messages to still be reindexed.
+
+**Note: For search reindex process, we create separate batches based on the search index name; this way we can insert documents to the correct index for both reindex jobs and publishing new content and data**
+
+**8: Multi-operational request to update/add docs to an index**
```

Comment:
why do we connect direct to ES here? just interested in the answer really.

Reply:
So the only other place that talks to ES is the Search API - but I generally see the search data pipeline as separate entity to the Search API and hence didn't want to use it. Other more minor reasons, the importer may change a lot and didn't want to be restricted by backwards compatibility or versioning by using a new endpoint on the search API; less latency. The search data pipeline is only hosted in publishing subnet, less chance of exposing ability for users to write to database - even though this is the situation via other DP APIs.

Reply:
Generally it seemed to make sense to do it this way